engine/storage/src/main/java/org/apache/cloudstack/storage/endpoint/DefaultEndPointSelector.java

```diff
@@ -128,9 +132,24 @@ protected EndPoint findEndPointInScope(Scope scope, String sqlBase, Long poolId)
             }
         }
 
+        List<Long> dedicatedHosts = dedicatedResourceDao.listAllHosts();
+
         // TODO: order by rand() is slow if there are lot of hosts
         sbuilder.append(") t where t.value<>'true' or t.value is null");    //Added for exclude cluster's subquery
-        sbuilder.append(" ORDER by rand() limit 1");
+        sbuilder.append(" ORDER by ");
+        if (dedicatedHosts.size() > 0) {
+            Collections.shuffle(dedicatedHosts); // Randomize dedicated hosts as well.
+            sbuilder.append("field(t.id, ");
+            Iterator<Long> hostIt = dedicatedHosts.iterator();
+            while (hostIt.hasNext()) { // put dedicated hosts at the end of the result set
```

Comment:
@Spaceman1984 the ordering based on the dedicated host ids here would put them on the top, can you confirm the result here?

Reply:
@sureshanaparti  I ran the SQL multiple times, the dedicated hosts were always at the end of the result set.

Reply:
> @sureshanaparti I ran the SQL multiple times, the dedicated hosts were always at the end of the result set.

Let's consider host ids: 3, 4 are dedicated, then the query does " _ORDER by field(t.id, '3','4'), rand() limit 1_". Am I correct?

Reply:
Yes @sureshanaparti

Reply:
thanks for the confirmation @Spaceman1984